### PR TITLE
6312 Лаврук

### PR DESCRIPTION
**Pull request by alxmcs**
Это часть кода дипломного проекта одного из моих студентов, написанная около 3х лет назад.
Здесь огромное количество code smells. Твоя задача - найти наибольшее их количество и предложить варианты от них избавиться.

### CODE REVIEW

File: baseFunctions.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    if (("room" in y1) and ("room" in y2)) and (("android" in y1) and ("android" in y2)): dec = 1
+    if (("room" in y1) and ("room" in y2)) and (("iphone" in y1) and ("iphone" in y2)): dec = 1
+    if (("electro" in y1) and ("electro" in y2)): dec = 1
+    if (("second_pink" in y1) and ("second_pink" in y2)): dec = 1
+
+    return dec
+
+def spectrogramm(x):
+    f1, t1, s1 = spectrogram(x)
+    return f1, t1, s1
```
Comment (Mlavrukk) [inline baseFunctions.py:41]: Неиспользуемые переменные, заменить на return spectrogram(х)

File: statisticsClassification.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+from sklearn.ensemble import RandomForestClassifier as classifier
+from sklearn.model_selection import train_test_split
+import pandas
+np.seterr(divide='ignore', invalid='ignore')
+
+data_set, files, srs = get_data_set("dataset")
+
+chi_points = []
+umann_points = []
+student_points = []
```
Comment (Mlavrukk) [inline statisticsClassification.py:14]: Несоответствие имен переменных и функций: chi_points, umann_points, student_points. Заменить на :  chi_squared_values
 mann_whitney_values
 t_test_values

File: statisticsClassification.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+labels = []
+middles = []
+all_vec = []
+
+train_len = round(len(data_set))
```
Comment (Mlavrukk) [inline statisticsClassification.py:19]: в коде используются константы без объяснения их значения. 
train_len = round(len(data_set)) 
заменить на :  train_len -> TRAIN_SIZE = 0.7

File: statisticsClassification.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+outside_points_chi = []
+outside_points_student = []
+outside_points_umann = []
+predicted = []
+
+chi_model = classifier()
+umann_model = classifier()
+stud_model = classifier()
+mid_model = classifier()
+all_model = classifier()
```
Comment (Mlavrukk) [inline statisticsClassification.py:52]: # Избыточность кода
chi_model = classifier()
umann_model = classifier()
stud_model = classifier()
mid_model = classifier()
all_model = classifier() 

 Заменить на: model = classifier().fit(all_x_train, all_y_train)

File: statisticsClassification.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+chi_points = np.array(chi_points)
+chi_points = chi_points.reshape(-1,1)
```
Comment (Mlavrukk) [inline statisticsClassification.py:55]: в коде повторяются одинаковые операции с разными переменными, например, chi_points = np.array(chi_points), chi_points = chi_points.reshape(-1,1). Это может привести к избыточности и сложности кода. 
Решение: использовать циклы или функции для упрощения кода:
 for points in [chi_points, umann_points, student_points, middles, all_vec]: points = np.array(points).reshape(-1,1)

File: statisticsClassification.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+middles = chi_points.reshape(-1,1)
+all_vec = np.array(all_vec)
+all_vec = chi_points.reshape(-1,1)
+
+
+chi_x_train, chi_x_test, chi_y_train, chi_y_test = train_test_split(chi_points, labels, test_size=0.30)
+umann_x_train, umann_x_test, umann_y_train, umann_y_test = train_test_split(umann_points, labels, test_size=0.30)
+stud_x_train, stud_x_test, stud_y_train, stud_y_test = train_test_split(student_points, labels, test_size=0.30)
+mid_x_train, mid_x_test, mid_y_train, mid_y_test = train_test_split(middles, labels, test_size=0.30)
+all_x_train, all_x_test, all_y_train, all_y_test = train_test_split(all_vec, labels, test_size=0.30)
```
Comment (Mlavrukk) [inline statisticsClassification.py:70]: в коде используются константы без объяснения их значения. 
test_size=0.30 # test_size -> TEST_SIZE = 1 - TRAIN_SIZE

File: statisticsClassification.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+middles = []
+all_vec = []
+
+train_len = round(len(data_set))
+
+train_split = data_set[0:]
+
+for i in range(len(train_split)):
+    for j in range(i + 1, len(train_split)):
+        [a, b, c] = statistics.use_stat_for_spectr(train_split[i], train_split[j])
```
Comment (Mlavrukk) [inline statisticsClassification.py:25]: [a, b, c] = statistics.use_stat_for_spectr(train_split[i], train_split[j]) 

Заменить на : a, b, c -> chi_squared, mann_whitney, t_test

File: statisticsClassification.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+report_uman = [accuracy_score(umann_y_test, umann_pred), f1_score(umann_y_test, umann_pred), precision_score(umann_y_test, umann_pred), recall_score(umann_y_test, umann_pred)]
+report_stud = [accuracy_score(stud_y_test, stud_pred), f1_score(stud_y_test, stud_pred), precision_score(stud_y_test, stud_pred), recall_score(stud_y_test, stud_pred)]
+report_mid = [accuracy_score(mid_y_test, middles_pred), f1_score(mid_y_test, middles_pred), precision_score(mid_y_test, middles_pred), recall_score(mid_y_test, middles_pred)]
+report_all = [accuracy_score(all_y_test, all_pred), f1_score(all_y_test, all_pred), precision_score(all_y_test, all_pred), recall_score(all_y_test, all_pred)]
+
+print("1", report_chi)
+print("2", report_uman)
+print("3", report_stud)
+print("4", report_mid)
+print("5", report_all)
```
Comment (alxmcs) [inline statisticsClassification.py:95]: Мне бы хотелось почитать в ревью предложения насчет того, как можно сократить дублирование этого кода  - здесь несколько блоков кода подряд с практически идентичным набором операций

File: statisticsClassification.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+chi_points = np.array(chi_points)
+chi_points = chi_points.reshape(-1,1)
+umann_points = np.array(umann_points)
+umann_points = chi_points.reshape(-1,1)
+student_points = np.array(student_points)
+student_points = chi_points.reshape(-1,1)
+middles = np.array(middles)
+middles = chi_points.reshape(-1,1)
+all_vec = np.array(all_vec)
+all_vec = chi_points.reshape(-1,1)
```
Comment (Mlavrukk) [inline statisticsClassification.py:63]: Можно использовать цикл for или while для повторения одинаковых операций над разными переменными: 
points_list = [chi_points, umann_points, student_points, middles, all_vec]
for points in points_list:
    points = np.array(points)
    points = points.reshape(-1,1)

File: statisticsClassification.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+
+chi_x_train, chi_x_test, chi_y_train, chi_y_test = train_test_split(chi_points, labels, test_size=0.30)
+umann_x_train, umann_x_test, umann_y_train, umann_y_test = train_test_split(umann_points, labels, test_size=0.30)
+stud_x_train, stud_x_test, stud_y_train, stud_y_test = train_test_split(student_points, labels, test_size=0.30)
+mid_x_train, mid_x_test, mid_y_train, mid_y_test = train_test_split(middles, labels, test_size=0.30)
+all_x_train, all_x_test, all_y_train, all_y_test = train_test_split(all_vec, labels, test_size=0.30)
+
```
Comment (Mlavrukk) [inline statisticsClassification.py:71]: Можно использовать функцию, которая принимает на вход разные параметры и возвращает нужные результаты:
def split_data(points, labels, test_size):
    x_train, x_test, y_train, y_test = train_test_split(points, labels, test_size=test_size)
    return x_train, x_test, y_train, y_test

chi_x_train, chi_x_test, chi_y_train, chi_y_test = split_data(chi_points, labels, 0.30)
umann_x_train, umann_x_test, umann_y_train, umann_y_test = split_data(umann_points, labels, 0.30)
stud_x_train, stud_x_test, stud_y_train, stud_y_test = split_data(student_points, labels, 0.30)
mid_x_train, mid_x_test, mid_y_train, mid_y_test = split_data(middles, labels, 0.30)
all_x_train, all_x_test, all_y_train, all_y_test = split_data(all_vec, labels, 0.30)

File: statisticsClassification.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+mid_model.fit(mid_x_train, mid_y_train)
+all_model.fit(X=all_x_train, y=all_y_train)
+
+
+chi_pred = chi_model.predict(chi_x_test)
+stud_pred = stud_model.predict(stud_x_test)
+umann_pred = umann_model.predict(umann_x_test)
+middles_pred = mid_model.predict(mid_x_test)
+all_pred = all_model.predict(all_x_test)
+
```
Comment (Mlavrukk) [inline statisticsClassification.py:84]: Использовать списки, словари или другие структуры данных, которые позволяют хранить и обрабатывать несколько значений под одним именем:
models = {}
predictions = {}
for name, points in zip(["chi", "umann", "stud", "mid", "all"], points_list):
    models[name] = classifier()
    x_train, x_test, y_train, y_test = split_data(points, labels, 0.30)
    models[name].fit(x_train, y_train)
    predictions[name] = models[name].predict(x_test)

**Review by alxmcs — COMMENTED** [review]
Зачтено
